tests/README.rst

```diff
@@ -0,0 +1,76 @@
+===========
+Don't Panic
+===========
+
+The Hypothesis test suite is large, but we've written these notes to help you
+out.  It's aimed at contributors (new and old!) who know they need to add tests
+*somewhere*, but aren't sure where - or maybe need some hints on what kinds of
+tests might be useful.  Others might just be interested in how a testing
+library tests itself!
+
+
+The very short version
+======================
+
+- To improve code coverage (eg because ``make check-coverage`` / the Travis
+  build is failing), go to ``cover/``
+- For longer / system / integration tests, look in ``nocover/``
+- For tests that require an optional dependency, look in the appropriate
+  directory
+
+In each case there will be further notes to help you.
+
+
+About each group of tests
+=========================
+
+Still here?  Here's a note on what to expect in each directory.
+
+``common/``
+    Useful shared testing code, including test setup and a few helper
+    functions in ``utils.py``.  Also read up on
+    `pytest <https://docs.pytest.org/en/latest/contents.html>`_
+    features such as ``mark.parametrize``, ``mark.skipif``, and ``raises``
+    for other functions that are often useful when writing tests.
+
+``cover/``
+    The home of enough tests to get 100% branch coverage - you'll probably
+    spend most of your time here.  Ideally this would be as quick as possible
+    while remaining comprehensive.  More details inside.
+
+``datetime/``
+    Tests for the deprecated ``hypothesis.extra.datetime`` module, which
+    depends on the ``pytz`` package.
+
+``django/``
+    Tests for the Django extra.  Includes a toy application, to give us lots
+    of models to generate.
+
+``fakefactory/``
+    *Tests for the deprecated Faker extra.*
+
+``nocover/``
+    More expensive and longer-running tests, typically used to test trickier
+    interactions or check for regressions in expensive bugs.  Lots of tests
+    about how values shrink, databases, compatibility, etc.
+
+``numpy/``
+    Tests for the Numpy extra.
+
+``pandas/``
+    Tests for the Pandas extra.
+
+``py2/``
+    Tests that require Python 2 - in order to check that it only allows
+    constructs that also work on Python 3.
+
+``py3/``
+    Tests that require Python 3.  Includes checking that unicode identifiers
+    and function annotations don't break anything, asyncio tests, and tests
+    for inference from type hints.
+
+``pytest/``
+    For integration between Hypothesis and pytest.
+
+``quality/``
+    Tests that various hard-to-find examples do in fact get found by Hypothesis.
```

Comment:
It's more than that - it also has a bunch of things about shrinking.  I'm not sure how to neatly summarise the unifying theme though - I guess it's mostly intended for tests of the form "Hypothesis finds an example of this condition" + assertions about which example it finds.